salesforce/metadataApi.go

```diff
@@ -0,0 +1,184 @@
+package salesforce
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/amp-labs/connectors/common"
+	"golang.org/x/oauth2"
+)
+
+const (
+	versionPrefix = "v"
+)
+
+var (
+	ErrCreateMetadata  = fmt.Errorf("error in CreateMetadata")
+	ErrCreatingRequest = fmt.Errorf("error in creating request")
+)
+
+func (c *Connector) CreateMetadata(ctx context.Context, operation *common.XMLData, tok *oauth2.Token) (string, error) {
+
+	req, err := c.prepareRequest(ctx, operation, tok)
+	if err != nil {
+		return "", err
+	}
+
+	res, body, err := c.makeRequest(ctx, req)
+	// below is a workaround to refresh token if it is expired
+	// normally oauth2 library should handle this
+	// but SOAP API does not take token in header
+	// but takes it in body
+	// So in case of 500 error and INVALID_SESSION_ID in body
+	// we know it is session expired, and automatically the client refreshes token
+	// tok object will be updated with new token automatically after failing first call
+	// we simply make another call with updated token.
+	if res.StatusCode == 500 && strings.Contains(string(body), "INVALID_SESSION_ID") {
+		req, err := c.prepareRequest(ctx, operation, tok)
+		if err != nil {
+			return "", errors.Join(ErrCreateMetadata, err)
+		}
+
+		res, body, err = c.makeRequest(ctx, req)
+		if err != nil {
+			return string(body), fmt.Errorf("%w: %s", ErrCreateMetadata, string(body))
+		}
+	}
+
+	if res.StatusCode < 200 || res.StatusCode > 299 {
+		return "", fmt.Errorf("%w: %s", ErrCreateMetadata, string(body))
+	}
+
+	return string(body), nil
+}
+
+func (c *Connector) prepareRequest(ctx context.Context, operation *common.XMLData, tok *oauth2.Token) (*http.Request, error) {
+	data := preparePayload(operation, tok.AccessToken)
+
+	endPointURL, err := url.JoinPath(c.Client.Base, "services/Soap/m/"+removePrefix(c.APIVersion(), versionPrefix))
```

Comment:
can we actually have a different constant called `APIVersionSOAP` or something like that, rather than having to remove the prefix like this, and needing to expose an `APIVersion` method. It also gives us the flexibility in the future to use a different version of the SOAP API vs the REST API.